### ISSUE TITLE
PHP7 & new memcached library

### ISSUE THREAD

**Issue by AlexKapustin** [facebook/mcrouter#143]
Greetings,

I've noticed problem that getStats method for new memcached library doesn't work.

```php
$m = new Memcached();
$m->addServer('localhost', 1111);
$a = $m->getStats();
var_dump($a);  // bool(false) in output
```
And there is the problem:
```getsockopt(4, SOL_SOCKET, SO_ERROR, [0], [4]) = 0
sendto(4, "version\r\n", 9, MSG_NOSIGNAL, NULL, 0) = 9
recvfrom(4, 0x7fdc0b833210, 8196, 16384, 0, 0) = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=4, events=POLLIN}], 1, 5000)  = 1 ([{fd=4, revents=POLLIN}])
recvfrom(4, "VERSION mcrouter 2.0.0\r\n", 8196, MSG_NOSIGNAL, NULL, NULL) = 24
shutdown(4, SHUT_RDWR)                  = 0
close(4)                                = 0
```

Looks like memcached library tries to get memcached version to determinate protocol, but as it returns not compatible data - fails.

Do you have any plan to fix this issue ? Maybe by adding some option for returned real memcached version ?

Have a nice day. 

Best Regards,
Alex

**Comment by sodabrew** [comment]
If you're seeing this problem, you might be using a very old version of mcrouter, or you might have compiled your own version with the MCROUTER_PACKAGE_STRING explicitly set to "mcrouter 2.0.0".

The issue you're describing was reported in #62 and fixed in a1e485e81ab03b390ece8d6728d87499160ea613

The configure script in master retains the version-number-first ordering with a comment to keep it this way for libmemcached compatibility:
https://github.com/facebook/mcrouter/blob/master/mcrouter/configure.ac#L12


... I spent some time researching the problem, so here are my notes for posterity ...


The PHP memcached extension uses the libmemcached C library. The PHP memcached extension calls libmemcached's `memcached_version()` function, which loops over the configured memcacheds (in this case, just your local mcrouter instance) and asks each memcached for its version number, storing that information within libmemcached for reference by other version-dependent functionality (i.e. getting stats, because the expected set of stats keys has changed over time). If any of the memcached servers (i.e. mcrouter) returns an expected VERSION string, the libmemcached returns an error from `memcached_version`. The PHP memcached extension returns from the `getStats()` method if the `memcached_version` function returns an error, before calling the `memcached_stats` function.

The underlying implementation is here: https://bazaar.launchpad.net/~tangent-trunk/libmemcached/1.0/view/head:/libmemcached/stats.cc#L840

Basically libmemcached is expecting a version of this format;
```
version
VERSION 1.4.22
```

But mcrouter is giving a version in this format:
```
version
VERSION mcrouter 2.0.0
```

As it happens, libmemcached will ignore any trailing text, so the following would be a valid reponse:
```
version
VERSION 2.0.0 mcrouter
```

... at this point, I went digging in the code, figured out that someone else had already figured this out, and wrote the new reply up at the top.
